lib/apiservers/engine/backends/container_proxy.go

```diff
@@ -771,72 +768,87 @@ func (c *ContainerProxy) AttachStreams(ctx context.Context, vc *viccontainer.Vic
 		}
 	}
 
+	var inCtx context.Context
+	var inCancel, cancel func()
+
 	if ca.UseStdin {
 		wg.Add(1)
+		inCtx, inCancel = context.WithCancel(ctx)
+		defer inCancel()
+	}
+
+	if ca.UseStdout {
+		wg.Add(1)
+		ctx, cancel = context.WithCancel(ctx)
+		defer cancel()
+	}
+
+	if ca.UseStderr {
+		wg.Add(1)
+		if cancel == nil {
+			ctx, cancel = context.WithCancel(ctx)
+			defer cancel()
+		}
+	}
+
+	if ca.UseStdin {
 		go func() {
 			defer wg.Done()
-			err := copyStdIn(ctx, plClient, vc, clStdin, keys)
+			err := copyStdIn(inCtx, plClient, vc, clStdin, keys)
 			if err != nil {
-				log.Errorf("container attach: stdin (%s): %s", vc.ContainerID, err.Error())
+				log.Errorf("container attach: stdin (%s): %s", vc.ContainerID, err)
 			} else {
-				log.Infof("container attach: stdin (%s) done: %s", vc.ContainerID)
+				log.Infof("container attach: stdin (%s) done", vc.ContainerID)
 			}
 
-			// no need to take action if we are canceled
-			// as that means error happened somewhere else
-			if ctx.Err() == context.Canceled {
-				log.Infof("returning from stdin as context canceled somewhere else")
-				return
+			// close stdout and stderr
+			if cancel != nil {
```

Comment:
In the case where stdin is closed but the container has not yet finished generating output this will truncate the remaining output. This cancellation should only apply during detach.